### PR TITLE
Add some small improvements to make the list load a bitfaster

### PR DESCRIPTION
**Pull request by goanpeca**
Add small improvement on perceived speed of loading:

- Use batches of 2 (bigger batches start making the UI block on my machine)
- Use 62 ms between batches
- Allow for searching of plugins not yet on list, and add them
- throttle filter keystrokes so a filter starts only after some miliseconds
- add (X/N) to both list when filtering so it is clear that X plugins are filtered out of a total of N
- Change search type based on search text length, so if only `start_with_char` or less chars are typed then we match with starting letters only, otherwise we see if the text is contained in name/description etc. We also search for plugins that start with `napari-` plus the number of characters defined by `start_with_char` 

Screenshot 2024-05-29 at 11 48 50 AM

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
<15 s to fully populate on my machine vs 45 on main!

File: napari_plugin_manager/qt_plugin_dialog.py
Comment (psobolewskiPhD) [inline napari_plugin_manager/qt_plugin_dialog.py:725]: Can you explain the rationale for this? It's noticeable as a lag when typing. I tried 100 and it felt better.

File: napari_plugin_manager/qt_plugin_dialog.py
Comment (psobolewskiPhD) [inline napari_plugin_manager/qt_plugin_dialog.py:1076]: Could we make the `2` a variable, like `batch_size` so it's better documented? and update the docstring (it's not one by one, but in batches)
Not sure it makes sense to make it a kwarg, but just so it's not a random integer.

For the record I tried 10 on my machine it made scrolling laggy--not good.
`3` felt pretty good though (2020 M1), but we should target the lower common denominator.

File: napari_plugin_manager/qt_plugin_dialog.py
Comment (psobolewskiPhD) [inline napari_plugin_manager/qt_plugin_dialog.py:729]:
```suggestion
        # Add items in batches with a pause to avoid blocking the UI
```

**Review by psobolewskiPhD — APPROVED** [review]
For me this makes a marked difference and with showing how many are to be loaded makes it easy to estimate how long to wait.
filtering works well, thought I don't love the 200 ms pause -- left a question about that.

I'm going to approve, but hopefully one more person takes a look before merging.

File: napari_plugin_manager/qt_plugin_dialog.py
Comment (goanpeca) [reply napari_plugin_manager/qt_plugin_dialog.py:725]: Hi @psobolewskiPhD,  I think between 100-150 ms is what is a expected time between keypresses, so to avoid triggering a filter for every keystroke when one wants to write a word.

I just bumped it a bit to 200, but 150 should work fine. 100 might be a little on the edge? And yes it is noticeable so maybe we should decrease it

I know it is noticeable, so maybe going for 150 feels better?

File: napari_plugin_manager/qt_plugin_dialog.py
Comment (goanpeca) [reply napari_plugin_manager/qt_plugin_dialog.py:1076]: Indeed, what you describe was exactly what I found, 2 is a sweet spot in this case.

**Comment by goanpeca** [comment]
Thanks a lot @psobolewskiPhD for the review. You raised valid points and indeed the 200 can be lowered. 

https://arxiv.org/pdf/1208.6109 so english average word is 5.1 letters, assuming a typing speed 60 word per minute (https://typing-speed.net/keystrokes-per-minute I tried this one and got 361 letters in 60 seconds, I am not a very fast typer hahah), so we can do some quick math and get:

5.1*60 + 60 = 366 letters in 60 seconds (the extra 60 is to account for spaces), so:
60 seconds / 366 letters =~ 0.163 seconds, which is pretty similar to what I got trying the other service.

So I think I have read somewhere that 150ms was a good number for interstroke time, I just bumped it a bit, but maybe 150 works after all :)

File: napari_plugin_manager/qt_plugin_dialog.py
Comment (psobolewskiPhD) [inline napari_plugin_manager/qt_plugin_dialog.py:725]:
```suggestion
	# timer to avoid triggering a filter for every keystroke
        self._filter_timer.setInterval(120)  # ms
```

**Comment by psobolewskiPhD** [comment]
Ah, thanks for the explanation. So you're sort of debouncing.
What I notice is a delay between the first character and 2nd -- when the list is still loading.
Open the manager and try typing `lung` into the filter, all one word, naturally.
L will appear and then a delay and the rest.

File: napari_plugin_manager/qt_plugin_dialog.py
Comment (psobolewskiPhD) [inline napari_plugin_manager/qt_plugin_dialog.py:893]:
```suggestion
```
This comment isn't correct (120 ms now) because of change above, so I made a suggestion to comment there instead and then we can remove this one.

**Comment by goanpeca** [comment]
Hi @psobolewskiPhD, yes, indeed debouncing basically, to avoid triggering the filter for just one letter, we could make sure to always have at least 2 letters or 3 before actually doing any filtering, that would also work.

Will take a look at superqt, I thought about it but did not test it.

Will check the first letter issue that you mention!

Thanks again for the comments and the review :)

**Comment by jaimergp** [comment]
> Will check the first letter issue that you mention!

I've dealt with this kind of issue in the past and resorted to the following strategy:

- If `len(query) == 1`, then we only show results _starting_ with that letter. I doubt folks want all packages containing `a` 🤷 
- Else, we show results containing it.

Depending on the volume of results for `1` (and `2`?), maybe you can adjust the `startsWith` / `contains` threshold differently.

If this is too weird, then I suggest disabling search for queries of length == 1.

**Comment by goanpeca** [comment]
It was fairly simple to implement your suggestion @jaimergp , so I think we are good now with this PR :)

**Comment by Czaki** [comment]
> If `len(query) == 1`, then we only show results _starting_ with that letter. I doubt folks want all packages containing `a` 🤷

It should also show packages starting from `napari-a` in our case.

**Comment by goanpeca** [comment]
>It should also show packages starting from napari-a in our case.

Done @Czaki !

**Comment by psobolewskiPhD** [comment]
Definitely feels better, but I still see a delay when typing a word like lung naturally, after the first letter. 
Shouldn't the 120 ms timer would prevent anything after just the first letter (debouncing), until I finish? It feels like it need to trigger -- off the first letter -- and the debounces the rest of them.

File: napari_plugin_manager/qt_plugin_dialog.py
Comment (jaimergp) [inline napari_plugin_manager/qt_plugin_dialog.py:704]:
```suggestion
            if len(text) <= starts_with_chars:
                flag = Qt.MatchFlag.MatchStartsWith
                queries = (text, f'napari-{text}')
            else:
                flag = Qt.MatchFlag.MatchContains
                queries = (text,)
            shown = {id(it) for query in queries for it in self.findItems(query, flag)}
```

What about this refactor? :)

Comment (goanpeca) [reply napari_plugin_manager/qt_plugin_dialog.py:704]: Not a fan of double nested comprehensions (it becomes a bit unreadable) but sure why not :)

**Comment by goanpeca** [comment]
>Definitely feels better, but I still see a delay when typing a word like lung naturally, after the first letter.
Shouldn't the 120 ms timer would prevent anything after just the first letter (debouncing), until I finish? It feels like it need to trigger -- off the first letter -- and the debounces the rest of them.

Will check again, as I am not seeing that on my side 🤔

**Comment by psobolewskiPhD** [comment]
Feels really solid now!
I do get a traceback when closing napari though:
```
Exception ignored in: <function CameraMonitor.__del__ at 0x1437a4180>
Traceback (most recent call last):
  File "/Users/piotrsobolewski/Dev/napari/napari/_qt/qt_viewer.py", line 1387, in __del__
    self._reset_camera_moving.cancel()
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/napari-dev/lib/python3.12/site-packages/superqt/utils/_throttler.py", line 291, in __get__
    parent = self.parent()
             ^^^^^^^^^^^^^
RuntimeError: wrapped C/C++ object of type ThrottledCallable has been deleted
```
(open the manager, search for something, close the manager, quit napari)

Also, I was surprised to see the search persist through closing the widget, but I guess that's the caching from the other PR.
Honestly, not sure what the "right" behavior is, so not a blocker at all. Resolving the RunetimeError though would be good.

**Comment by goanpeca** [comment]
Hi @psobolewskiPhD thanks a lot for all the feedback. Hmmm weird, that traceback feels unrelated. But I will try to reproduce and see if soemthing i did messd it up!

>Also, I was surprised to see the search persist through closing the widget, but I guess that's the caching from the other PR. Honestly, not sure what the "right" behavior is, so not a blocker at all. Resolving the RunetimeError though would be good.

Indeed! I can reset the search on reopen, that would probably feel more natural. Also working on a separate PR to add notifications when things ended or finished if for example you started installing something and closed the dialog. That way the user knows when something finished installing or if an error occured ✌🏼

**Comment by psobolewskiPhD** [comment]
Shoot hang on, I could have had a branch checked out in napari repo. When I had this checked out. 😬

**Review by psobolewskiPhD — APPROVED** [review]

**Comment by goanpeca** [comment]
@psobolewskiPhD now the filter gets cleared on hide/close to match your expectations, also added a small fix and now is good to go. Will merge after tests pass so I can focus https://github.com/napari/napari-plugin-manager/pull/45 which is the final piece of work !

**Comment by goanpeca** [comment]
The failing test does not seem reproducible. Tested locally, will merge an iterate on the last PR

**Pull request merged**
